src/main/kotlin/org/wfanet/measurement/duchy/daemon/mill/liquidlegionsv2/LiquidLegionsV2Mill.kt

```diff
@@ -533,45 +535,49 @@ class LiquidLegionsV2Mill(
     return completeComputation(nextToken, CompletedReason.SUCCEEDED)
   }
 
+  // Gets the order of all duchies in the MPC ring. Since it is a ring, we put the aggregator at the
+  // end for simplicity.
+  private fun getMpcDuchyRing(token: ComputationToken): List<String> {
+    val nonAggregatorsSet = cryptoKeySet.allDuchyPublicKeys
+      .filter { it.key != aggregatorId }
+      .map { Duchy(it.key, it.value.element.toStringUtf8()) }.toSet()
+    return getDuchyOrderByPublicKeysAndComputationId(
+      nonAggregatorsSet, token.globalComputationId
+    ) + aggregatorId
+  }
+
   private fun nextDuchyStub(token: ComputationToken): ComputationControlCoroutineStub {
-    val nextDuchy = token.computationDetails.liquidLegionsV2.outgoingNodeId
+    val mpcDuchyRing = getMpcDuchyRing(token)
+    val nextDuchy = getNextDuchy(mpcDuchyRing, duchyId)
     return workerStubs[nextDuchy]
       ?: throw PermanentComputationError(
         IllegalArgumentException("No ComputationControlService stub for next duchy '$nextDuchy'")
       )
   }
 
-  private fun aggregatorDuchyStub(token: ComputationToken): ComputationControlCoroutineStub {
-    val aggregatorDuchy = token.computationDetails.liquidLegionsV2.aggregatorNodeId
-    return workerStubs[aggregatorDuchy]
+  private fun aggregatorDuchyStub(): ComputationControlCoroutineStub {
+    return workerStubs[aggregatorId]
       ?: throw PermanentComputationError(
         IllegalArgumentException(
-          "No ComputationControlService stub for primary duchy '$aggregatorDuchy'"
+          "No ComputationControlService stub for primary duchy '$aggregatorId'"
```

Comment:
Nit: Primary is old term for aggregator.

Reply:
Done